### DISCUSSION TITLE
dhart performance & bottlenecks

### DISCUSSION THREAD

**Discussion by mariuszhermansdorfer** [cadop/dhart#33] [General]
As mentioned in #29, let's discuss various performance improvements strategies for dhart.

**Comment by cadop** [comment]
One concern is the time it takes to pass data from C# to the C interface.  It may be helpful if you can try the two different methods of generating rays, one where you loop over the directions and one where you make two lists for directions and origins and don't iterate at all in c#. 

Another idea for trying to understand the performance is to generate all of the origins and the direction to be outside the model bounding box, each ray would instantly fail the occlusion test and wouldn't bother traversing the bvh.  I think this might capture the data handling times in the interface.

**Reply by mariuszhermansdorfer** [reply]
Here are the test results.

500K origins, one ray -> Single Ray
500K origins, 500k rays -> Multiple Rays
![image](https://user-images.githubusercontent.com/49192999/203639409-8798a6ad-36a4-42dd-8b05-5e460aa5c532.png)


Moving the ray origins outside of the bounding box:
![image](https://user-images.githubusercontent.com/49192999/203639477-48edb8ad-08a9-4720-99fb-cc3251cf7a4a.png)

These results are benchmarked on a different laptop running on battery hence absolute values shouldn't be compared with the ones I previously posted. Having said that, the relative performance differences should still be correct. Multiple origins & single ray seems to be consistently faster than providing 2 arrays with matching length for origins and directions.

If your assumption is true, and moving ray origin points outside of the model bounding box exposes the data handling overhead, there seems to be some room for optimization here. These constitute roughly 50% of the total time.

To be clear, I measure these with a SystemDiagnostics.Stopwatch and only for one Method:
``` c#
EmbreeRaytracer.IntersectOccluded(bvh, analysisPoints, new Vector3D[] { sunDirections[0] }); 
```

**Reply by mariuszhermansdorfer** [reply]
One low hanging fruit could be to create an overload for this method which takes a flat array of floats as inputs instead of arrays of Vector3D:
https://github.com/cadop/dhart/blob/d7eefe6abcb2fce6e0f631665007da10c15b2ded/src/Csharp/packages/raytracer/src/RayTracerNative.cs#L309

Currently, we have to convert twice. First from Rhino’s Vector3d/Point3d to dhart’s Vector3D and then, within the method to float[].

**Reply by cadop** [reply]
--------------------------------
to clarify the multiple rays, the idea is to compare against your loop in c sharp. so something like:

```
make origin array of size 500_000 for each point
make direction array of the same vector with size 500_000
for each timestamp, 
- add the same 500_000 points to the origin array
- add the new directions of 500_000
```

The idea is, instead of c# passing 500k points 10 times, you pass 5mil points once.  However to do this, we need to map every point to a direction, so you need a 5 million point origin array with a 5 million vector direction array.   This should give the same result as using a 500k point origin array 10 times, with each time passing a single direction. So the timer would be across the loop of 10 calls to dhart, and the single 5milx5mil call.

but, i'm not sure this helps if its the data transfer

**Reply by cadop** [reply]
okay you are too fast in the responses! 

I'm not sure I understand if it will get the speed we want, you are already testing just the call itself right, not the extra conversion you are doing?

**Comment by cadop** [comment]
it seems you are pretty capable on the C# side, so give me some time and I am going to clarify the build instructions so you can edit the c# interface and build the API.  I'll still handle the embree interface, but this way you can test out any of these data conversion ideas quicker than posting on github :D

**Reply by mariuszhermansdorfer** [reply]
I will test the 5 mil origins & 5 mil directions idea. Still found it interesting to understand which if the two methods yields better results for a single sun direction.

You understand it right - I am only benchmarking the call. But the Vector3D -> float[] conversion is part of the call itself. And since we need to convert from Rhino’s data structures anyway, we can do it directly to a flat float array. You’re right, though, this will shave off only a couple ms. 

Good suggestion with me testing these on my own rather than typing it all out here :)

**Reply by cadop** [reply]
I updated the BUILD.MD to include some walkthrough images using Visual Studio.   Please open an issue if you run into any problems or need clarity!

https://github.com/cadop/dhart/blob/main/BUILD.md

**Comment by cadop** [comment]
@mariuszhermansdorfer I made a hacked test branch for you. https://github.com/cadop/dhart/tree/support-for-stream-of-rays-in-embree

You can take a look at the commits to see how its done, but the main points are:
- The parallelization is removed to simplify the test (its not clear to me yet how to efficiently split the data)
- One version uses a non-parallel version of the current dhart method and the other uses the occlusion stream with coherent flag
- There is a unittest to compare the versions
- You can build the project for csharp release and compare performance in your own application. To switch between the two, set `max_distance` to `99999`

I just copied the ray arrays, so its possible there is some other optimization going on (which is why it may be useful for you to test as end-to-end).  

If you follow the updated BUILD.MD instructions and install the DebugRelease config, there will be an out-install folder that has `HFUnitTests.exe` and you can run the unit tests (not necessary, but this is why the unittest example looks like).  This is the result I got for 1 million rays (using C++, not going through the C interface):

![image](https://user-images.githubusercontent.com/11399119/203663650-88deac43-0140-4c62-a3e8-5013da36768e.png)
[bin.zip](https://github.com/cadop/dhart/files/10079825/bin.zip)
Attached is a zip of the bin as well, hopefully that works for you without needing to build.

**Reply by mariuszhermansdorfer** [reply]
Thanks a lot for this! I tried building the repo today but ran into some issues and didn’t have time to troubleshoot it. Will report tomorrow on any progress.

Your results look promising, though! 15% speed up is definitely non-negligible.

I’m also thinking of implementing other optimization strategies:
1. Identifying origin points inside of buildings and excluding them before any rays are cast. In a typical urban environment, buildings can take up 30-50% of the total area. Excluding them would dramatically reduce the amount of intersection tests. Do you have any ideas on how to reuse the bvh to achieve that?
2. Instead of shooting all 500k rays at once, shoot a random subset and let the analysis converge to a final state over time. Similar to how path traced rendering works, this provides users with immediate but noisy feedback which gradually clears up. 
3. Same principle but applied to sun directions. After each hour is analyzed, results are displayed immediately and get refined as we iterate over the user-defined time span for the sunlight study.

**Reply by mariuszhermansdorfer** [reply]
> Attached is a zip of the bin as well, hopefully that works for you without needing to build.

I used the compiled DLLs from your previous post. The initial tests seem to be even better than what you saw on the c++ side - a speed up of 30-40% for a stream of coherent rays versus casting them individually. 

500k origins, 1 ray:
![image](https://user-images.githubusercontent.com/49192999/203925638-3227077d-a048-4999-ae76-39e98d383b6f.png)

500k origins, 11 rays:
![image](https://user-images.githubusercontent.com/49192999/203927251-4d48277f-71f9-4c33-9587-441eedc3c7a3.png)

500k origins, 100 rays:
![image](https://user-images.githubusercontent.com/49192999/203927413-c8888dce-32bf-4d45-b700-4b9eae84fca0.png)

If these gains can be translated to the parallel version, it would make a significant difference!

**Reply by mariuszhermansdorfer** [reply]
@cadop, I've done some further testing and - as mentioned earlier - there is a low hanging fruit for some minor optimization. Currently, we need to convert Rhino's `Point3d` and `Vector3d` to dhart's `Vector3D` which then gets converted to a `float[]`. 

https://github.com/cadop/dhart/blob/8aa1655a64fe4507f72687bc971a1a19aff286d6/src/Csharp/packages/raytracer/src/RayTracerNative.cs#L327-L328

This conversion is relatively slow:

https://github.com/cadop/dhart/blob/8aa1655a64fe4507f72687bc971a1a19aff286d6/src/Csharp/packages/base/src/NativeUtils.cs#L137-L161

Ideally, we would have an overload accepting `float[]` as arguments, rather than `Vector3D`
https://github.com/cadop/dhart/blob/8aa1655a64fe4507f72687bc971a1a19aff286d6/src/Csharp/packages/raytracer/src/RayTracerNative.cs#L309-L314


Then, we'd be able to convert Rhino's `Point3d` and `Vector3d` directly to a `float[]` like so:

``` c#

// Use Reflection to get an array of Point3d from a list 
// This is much faster than calling .ToArray() on a list

            Point3d[] points = (Point3d[])typeof(List<Point3d>)
                   .GetField("_items", BindingFlags.NonPublic | BindingFlags.Instance)
                   .GetValue(_analysisPoints);

// Use Unsafe.Add to iterate over the array and cast doubles to floats

            float[] flatFloatArray = new float[points.Length * 3];

            ref var s0 = ref points[0];
            for (int i = 0, j = 0; i < points.Length; i++, j++)
            {
                flatFloatArray[j] = (float)Unsafe.Add(ref s0, i).X;
                flatFloatArray[j++] = (float)Unsafe.Add(ref s0, i).Y;
                flatFloatArray[j++] = (float)Unsafe.Add(ref s0, i).Z;
            }
```

For 500K origin points and 11 direction vectors, we can reduce the conversion time by 80-90%:
![image](https://user-images.githubusercontent.com/49192999/204094934-88317c30-97ba-452e-8e8e-daedf8c6b5b6.png)
![image](https://user-images.githubusercontent.com/49192999/204095928-cfbd0025-e006-4d56-8929-d2d515617d2e.png)

**Reply by mariuszhermansdorfer** [reply]
Independently of adding another overload to the `C_CastOcclusionRays` method, you could replace the current version of `FlattenVectorArray(IEnumerable<Vector3D> vectors)` with this one:

``` c#
        public static float[] FlattenVectorArrayUnsafe(Vector3D[] vectors)
        {
            float[] out_array = new float[vectors.Length * 3];
            
            ref var s0 = ref vectors[0];
            for (int i = 0, j = 0; i < vectors.Length; i++, j++)
            {
                out_array[j] = (float)Unsafe.Add(ref s0, i).x;
                out_array[j++] = (float)Unsafe.Add(ref s0, i).y;
                out_array[j++] = (float)Unsafe.Add(ref s0, i).z;
            }

            return out_array;
        }
```

It should help with other methods relying on the Vector3D struct.

**Reply by cadop** [reply]
Hey @mariuszhermansdorfer Thanks for digging into this.  Were you able to get dhart to compile from source?  if not, could you let me know where it fails (maybe in a new discussion) so I can update the instructions.  Likewise, if you did get it to work what was the issue? Then I can update FAQ.  

As for checking rays inside of buildings vs. outside.  Depending on your configuration, this is a perfect application of the graph generator!   You can see a similar application in this paper https://dl.acm.org/doi/abs/10.5555/3465085.3465153  . The idea is, you generate a graph with a start point in the road/environment.  It will generate all nodes in a grid given your settings, and if there is no door/open entrance to a building (e.g. in a mass model, or all doors are closed), it won't generate nodes in those locations.  Then you can take the nodes from the graph and pass them to the occlusion raycast function.  The graph generator is pretty fast compared to other methods of getting nodes in 3d environments, but may be a bit slower than methods that rely on 2d grids.  

For your performance improvements, its possible you are doing tests on a newer CPU than me, and some architecture improvements have helped.  Also I used a single plane in my test, so perhaps as the bvh grows we see much more improvement.  I can change that function to always use streaming and coherent rays, since I realized it is guaranteed to be coherent when there is multiple origins with only one direction. 

Please make a PR for your C# interface changes/overloads, and if possible add some unittests (i'll do it if you don't want) so I can check it won't break anything else.  It's also a better workflow to give you credit, instead of me copying your code snippets and committing the changes myself :)

**Reply by cadop** [reply]
I have some python tutorials on this as well (unfortunately I don't think these tutorials exist for C#, perhaps in the future).  This is generating a graph and calculating a visibility graph https://cadop.github.io/dhart/Python%20Docs/build/html/examples/visibility/basic.html .  I think its similar, basically the empty regions are extruded volumes 'up'.  If you pull the repo with gitlfs you can see the models in this file https://github.com/cadop/dhart/blob/main/src/Python/dhart/Example%20Models/VisibilityTestCases.obj  It only generates the graph given a certain 'step height', so the blocks are excluded.

**Reply by mariuszhermansdorfer** [reply]
Thanks @cadop! I will look into the graph generator and try to wrap my head around how it works.

For now, I've made some good progress with implementing the overload mentioned above and faster casting logic. I was able to build the repo and test it from Rhino. The results are satisfying:

500k origins, 1 ray:
![image](https://user-images.githubusercontent.com/49192999/204162308-36ee8aaa-67dc-4acf-a686-272e42c1c977.png) previously: ![image](https://user-images.githubusercontent.com/49192999/203925638-3227077d-a048-4999-ae76-39e98d383b6f.png)

500k origins, 11 rays:
![image](https://user-images.githubusercontent.com/49192999/204162376-ee5c0e82-2ab6-46f2-9ce0-16edee7b512c.png) previously: ![image](https://user-images.githubusercontent.com/49192999/203927251-4d48277f-71f9-4c33-9587-441eedc3c7a3.png)

500k origins, 100 rays:
![image](https://user-images.githubusercontent.com/49192999/204162383-b3fd8395-ebe9-4482-be5d-517e4e4ab7b5.png) previously: ![image](https://user-images.githubusercontent.com/49192999/203927413-c8888dce-32bf-4d45-b700-4b9eae84fca0.png)

It seems that the data conversion step was taking about 6-8 ms per each call. Switching `CastOcclusionRays` to work with `float[]` removes this overhead.


Moving the bvh far away from vector origin points gives us this:
![image](https://user-images.githubusercontent.com/49192999/204163598-50ffc64a-a689-449f-83a9-2e506ab88737.png)
It seems that there is still some overhead in data piping between c# and c++.

**Comment by cadop** [comment]
@mariuszhermansdorfer The graph generator is explained in detail https://arxiv.org/pdf/2101.10503.pdf (figure 9 is a decent summary). 

The gist is; pick a start point which is a valid location you want to create a grid from.  It then does a breadth-first search of the environment to find the available edges in a grid. If there is something blocking the edge (like the wall of a building), it doesn't create it.  You can also set things like max slopes and steps (i.e. if the slope is too high it won't create an edge or new node).

**Comment by cadop** [comment]
@mariuszhermansdorfer Parallelized Stream method is implemented on latest commit of the stream branch.  I left the non-stream version with the 99999 flag for you to test again. Also please check that its giving the same results just in case there is a race condition.

**Reply by mariuszhermansdorfer** [reply]
Hm, I'm getting some interesting results here. It seems that the parallel version is actually slower than the sequential. It definitely spins all my cores but is twice as slow as the previous stream implementation:

![image](https://user-images.githubusercontent.com/49192999/204370818-7148a3d5-8c65-4bc4-826b-21b41afc72ab.png) previously: ![image](https://user-images.githubusercontent.com/49192999/204364994-148b2822-1e53-4d05-989d-fcee83f2a5dc.png)


![image](https://user-images.githubusercontent.com/49192999/204370665-fdfb4974-c347-432c-8960-717a4076da27.png)
 
Do your tests show similar results?

**Reply by cadop** [reply]
what about the non-stream version?

**Reply by cadop** [reply]
the fact its 36ms with single core could also be pointing out the issue; the overhead of parallelization is just not worth it with that size. You could try bumping to 10 million and see if it improves in comparison to the older single core method.  

... or i could just have an issue with my implementation

**Reply by mariuszhermansdorfer** [reply]
Do you mean this one?
https://github.com/cadop/dhart/blob/dfd18eadfe268bb623856796e6876818623d9e60/src/Cpp/raytracer/src/embree_raytracer.cpp#L645-L656

Very similar results:
![image](https://user-images.githubusercontent.com/49192999/204372053-fc12d6bc-5eb9-43ae-b0dd-3d993f07899d.png)

**Reply by cadop** [reply]
Yea that is the comparison you were doing here right?

![image](https://user-images.githubusercontent.com/11399119/204372310-bb906578-b1f1-457d-9e9b-b1d8ec55dc8d.png)

Something else might be going on, since thats almost the same value.

**Reply by cadop** [reply]
on the C++ side I am getting good results (two different runs): 

![image](https://user-images.githubusercontent.com/11399119/204372827-041ee8fd-1f17-4ac3-8e18-52058fa9cbac.png)


![image](https://user-images.githubusercontent.com/11399119/204372877-55bb57b9-fdf8-4f9f-ba1b-c892f38b36d6.png)


although I am using 1 mil rays, and my desktop has 20 cores/ 40 threads, so it runs much slower on single core, I usually get much better benefits from parallelization, so you also testing is very helpful

**Reply by mariuszhermansdorfer** [reply]
I will test with more rays and get back to you. 

Another idea I had was to simplify another data conversion on the c++ side. What is the rationale behind the conversions in lines 273 & 274?
https://github.com/cadop/dhart/blob/30648be3f7b9df812a61d0b9fee22f60eaf42f78/src/Cinterface/raytracer_C.cpp#L271-L279

In the end, we need float values anyway:
https://github.com/cadop/dhart/blob/30648be3f7b9df812a61d0b9fee22f60eaf42f78/src/Cpp/raytracer/src/embree_raytracer.cpp#L621-L626 

If it's only to make looping over data more convenient, then it could be restructured to operating directly on a pointer to the float array we pass from c#. This could shave off another few ms per call.

**Reply by cadop** [reply]
it would be better to discuss this in a new issue + discussion if needed.  There are a bunch of interface constraints, not least of which is that the C interface shouldn't rely on pointers managed by C#.  

like before, theres always a balance of clarity and performance. clarity also means its easier to maintain and extend the API to dhart.

**Reply by mariuszhermansdorfer** [reply]
OK, tested with 1M origin points and ran it in a loop for 100 different rays:

parallel stream: 
![image](https://user-images.githubusercontent.com/49192999/204392694-cc869e91-2e80-41a7-b5fe-1a04efaa1f12.png)
sequential stream: 
![image](https://user-images.githubusercontent.com/49192999/204392878-693fe4e6-c938-49a8-ba95-fe80c53db115.png)

For whatever reason sequential stream execution seems to be the fastest here.

**Reply by cadop** [reply]
I don't think the loop will make much difference, but what was the result with the parallel non-stream (which is the latest one that doesnt use stream ... max_distance!=99999)?

**Reply by mariuszhermansdorfer** [reply]
The results were very similar to the parallel-stream. Almost no difference here.

**Reply by mariuszhermansdorfer** [reply]
Just as a reminder, this is what we started with:
500k origins, 1 direction
![image](https://user-images.githubusercontent.com/49192999/204395058-36137752-cee6-41e6-b03a-7a47af902b49.png)

And this is the fastest configuration on my machine (sequential stream of rays + skip Vector3D and pass float[]
![image](https://user-images.githubusercontent.com/49192999/204395267-4f3a4f9e-8a94-4d2a-bd31-8a3f6ff4d2ee.png)


Down from approx. 60 ms for data conversion & raycasting to <40 ms. That's a 33% improvement!

**Reply by cadop** [reply]
Hm, I'll check tomorrow by putting all the combinations in the same dll.  Can you put the rhino and gh files on your branch so I can reproduce your testing.

**Comment by cadop** [comment]
I modified some parameters.  Could you check running the same function with the following keys for max_distance:

`
99996
99997
99998
99999
`
Here is the dll

[bin.zip](https://github.com/cadop/dhart/files/10115998/bin.zip)

One of them is giving me very good results, but I could have done something like "p-hacking". If it worked, it will be noticeable.

**Reply by mariuszhermansdorfer** [reply]
Ulala, 99997 is a clear winner over here!

500K origins, one direction:
![image](https://user-images.githubusercontent.com/49192999/204640062-37c738ff-53ab-4dcf-8220-889ff4224ee1.png)

500K origins, 11 directions:
![image](https://user-images.githubusercontent.com/49192999/204640154-144585cd-f3c4-43f5-88bf-fcfa40bf8f89.png)

How did you set it up?

Also, this is still using Vector3D on the c# side. Switching over to float[] would shave off another few ms.   
I singled out 99997 now, and it parallelizes nicely:
![image](https://user-images.githubusercontent.com/49192999/204642517-ddd289d8-381d-45e2-8c9e-05babd97e136.png)

**Reply by cadop** [reply]
wonderful! i was testing my earlier suspicion on the parallel overhead being the reason for the slowdown.  I changed the chunks for openmp to reduce overhead, which made the parallel non-streaming version much faster. 

I have to do some more tests for small numbers of rays before I merge into the main branch, but will push to the stream branch soon. Might need to do some checks for load balancing thresholds, but I think having the parallel flag could help.

given the performance improvement on this version (on my 20 core CPU I am getting 1 million rays in _4ms_ in c++), i'm going to put the streaming method on the backburner.  It has its own issues that will be alot more work for me to do safely and efficiently.

**Reply by mariuszhermansdorfer** [reply]
Makes a lot of sense to go with the parallel non-stream version if it is already this fast!

I’ll rebase the Grasshopper branch to the stream one, once you’ve pushed your changes there.

**Reply by cadop** [reply]
Its currently on the streaming branch if you had some near-term need for that extra speed, but there is no API change so it shouldn't matter for your modifications

**Reply by mariuszhermansdorfer** [reply]
I ported my previous optimizations to the streaming branch and created a new PR #41

Here are the results:
500K origins, 1 direction:  
![image](https://user-images.githubusercontent.com/49192999/205075252-d26db9bd-54b8-41b4-aaba-4bdf11011410.png), previously: ![image](https://user-images.githubusercontent.com/49192999/205075706-f7ad977c-1669-46fd-9c67-0c9db1ec1108.png)

500K origins, 11 directions:
![image](https://user-images.githubusercontent.com/49192999/205075886-e39dff4c-1e86-47e5-85db-07f16712cf7b.png), previously: ![image](https://user-images.githubusercontent.com/49192999/205076009-3f3af951-e5eb-4635-aa62-368fd4b9023e.png)

That's almost 50% improvement for only one direction (including data conversion times) and approx. 35-40% improvement for multiple directions.

**Reply by mariuszhermansdorfer** [reply]
Tested it on a newer laptop with 16 cores:
![image](https://user-images.githubusercontent.com/49192999/205301845-e6536e61-82c3-44cd-a643-de13eeb321cd.png)

500K origins, 1 direction:
![image](https://user-images.githubusercontent.com/49192999/205301861-dd3506b6-58fc-4618-ad9c-5cc401c1b1bd.png)

500K origins, 11 directions:
![image](https://user-images.githubusercontent.com/49192999/205302008-725040e3-0dbc-468a-967c-9bab7798cd1c.png)

I think we can easily call this a success! Thanks a lot for your openness to suggestions and rapid implementation of feature requests!

**Reply by cadop** [reply]
awesome!  Thanks for your suggestions and quick testing of them!

I will work through adding unit tests and checking that `unsafe` doesn't break things in Unity.

**Reply by mariuszhermansdorfer** [reply]
I found a bug in my unsafe looping algorithm. Will push an update later today, but it has to do with how I was indexing it. Here is the correct way:
![image](https://user-images.githubusercontent.com/49192999/205318059-847ad19f-2610-4c01-965b-0b22ff401135.png)
